src/pathfinding_igvc/src/AStar.cpp

```diff
@@ -6,10 +6,224 @@
  */
 
 #include <AStar.h>
+using namespace std;
+
+AStar::AStar(nav_msgs::OccupancyGrid occupancy_grid, GridPoint start,
+        GridPoint goal) {
+    this->_num_cols = occupancy_grid.info.width;
+    this->_num_rows = occupancy_grid.info.height;
+    this->_goal = goal;
+    this->_start = start;
+    this->_grid = occupancy_grid.data;
+
+    this->_cell_details = new CellDetail*[this->_num_rows];
+    for (int i = 0; i < this->_num_rows; i++) {
+        this->_cell_details[i] = new CellDetail[this->_num_cols];
+    }
+
+    /*
+     Create an open list
+     */
+    this->_open_list = set<GridPointWithScore>();
+
+    // Create a closed list and initialise it to false which means
+    // that no cell has been included yet
+    // This closed list is implemented as a boolean 2D array
+    this->_closed_list = new bool*[this->_num_rows];
+    for (int i = 0; i < this->_num_rows; i++) {
+        this->_closed_list[i] = new bool[this->_num_cols];
+    }
+}
 
 std::stack<AStar::GridPoint> AStar::run(nav_msgs::OccupancyGrid occupancy_grid,
                                         GridPoint start,
                                         GridPoint goal) {
-    // TODO Part 2
+    return AStar(occupancy_grid, start, goal).aStarSearch();
+}
```

Comment:
This interface is a bit confusing for me. Why don't we just use the Constructor and then call `aStarSearch` on the object (basically what you're doing here, just not sure why we have a function for it and a private constructor)?

Reply:
If we were to make the constructor public, that means that we can create an `AStar` object and reuse it for finding path multiple times. First of all, I don't see any use to create an `AStar` object and find the path more than once with the exact same occupancy grid and starting/goal point. Second of all, even if we were to reuse the `AStar` object, we would need to ensure that all the instance variables in this class are reinitialized. If we just make the constructor private, we don't need to worry about the `AStar` object being created and reused multiple times without all of its instance variables being reinitialized. 

I actually had this issue with DBSCAN and the bug was discovered pretty late since it only happened after the first callback. So making it private would avoid any issue of this sort

Let me know what you think/if you have any better suggestions!